RFCs/FS-1061-Allow-intrinsic-type-extensions-for-provided-types.md

```diff
@@ -0,0 +1,122 @@
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this change comes from the inherant limitations with the way that provided types can be used.  With C# code generation it's very common for the generated code to be produced as a [partial class](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods) where the user is free to extend the type in another file by using the partial class syntax, in F# this is not possible as it has no concept of a partial class.  F# does have intrinsic and optional type extensions, it makes sense to allow this syntax to be used on provided types to allow the generated code more customisation by the user.  Frameworks such as Xamarin.iOS require certain methods to be implemented as overloads which rules out using optional type extensions, indeed its common for frameworks to have these kind of constraints as C# is known to use partial classes for these situations.  Having intrinsic type extension syntax greatly simplifys the construction on these types.  
+
+# Detailed design
+[design]: #detailed-design
+
+The compiler needs to be modified so that instead of an error being generated when a provided type is extended the type is extended:
+
+```
+Error FS0192: internal error: FindTypeDefBuilder: <Method> not found (FS0192) (<Application>)
+```
+
+Instead the type is extended in the same way as a normal type extension.
+
+The proposed systax is the same as the existing intrinsic type extension syntax:
+
+Example code short form:
+
+```fsharp
+type t = BasicProvider.Generative.MyType with
+    member x.Foo = 42
+```
+
+Example code long form:
+
+```fsharp
+type t = BasicProvider.Generative.MyType
+
+type t with 
+    member x.Foo = 42
+```
+
+Given that `MyType` is just a plain type with a default constructor the resulting generated code wold look like this:
+```
+type t =
+  class
+    new : unit -> t
+    member Foo : int
+```
+
+An addition change to languge specification also needs to be changed so that an override can be applied as an intrinsic type extension too.   Imagine that BasicProvider.Generative.MyType has an abstract method as follows:
+
+```fsharp
+abstract member Rotate: float -> unit
+```
+
+It would be natural to describe this as follows:
+
+```fsharp
+type t = BasicProvider.Generative.MyType
+
+type t with 
+    member x.Foo = 42
+    override x.Rotate _ = ()
+```
+
+This would result in the following warning:
+```
+[FS0060] Override implementations in augmentations are now deprecated. Override implementations should be given as part of the initial declaration of a type.
+```
+
+So the final part of this  proposal is to allow override implementations in provided types **only**, as it is not always possible to provide the implementation as part of the initial declaration of the type especially if you are only the consumer of the type provider. 
```

Comment:
Nit: implementations in provided types for generative type providers only.

Reply:
Could you provide a concrete example of this being a limitation today? I think I understand why this is a limitation, but for the sake of documenting this relaxation of the deprecation for one specific case, an example would be helpful for others who go through and read this as the canonical source of why a change was made.

Reply:
done